src/realm/dictionary.cpp

```diff
@@ -508,17 +629,56 @@ std::pair<Dictionary::Iterator, bool> Dictionary::insert(Mixed key, Mixed value)
     }
 
     auto hash = key.hash();
-    ObjKey k(int64_t(hash & 0x7FFFFFFFFFFFFFFF));
+    ObjKey k(int64_t(hash & s_hash_mask));
 
     bool old_entry = false;
-    ClusterNode::State state;
-    try {
-        // We assume that we will most likely insert new values, so we try this first
+    ClusterNode::State state = m_clusters->try_get(k);
+    if (state.index == realm::npos) {
+        // key does not already exist
         state = m_clusters->insert(k, key, value);
     }
-    catch (const KeyAlreadyUsed&) {
-        state = m_clusters->get(k);
-        old_entry = true;
+    else {
+        // Check if key matches
+        if (do_get_key(state).compare_signed(key) == 0) {
+            old_entry = true;
+        }
+        else {
+            // We have a collision
+            m_clusters->create_collision_column();
+
+            if (m_clusters->find_sibling(state, key)) {
+                old_entry = true;
+            }
+            else {
+                // Create a key with upper bit set and link it
+                ObjKey alternative_key = k.get_unresolved();
```

Comment:
I think that getting a key with the upper bit set basically limits us to 1 collision per hash value. In my local test where I make `Mixed::hash()` into the worst possible hash function, the collision tests you added crash. If I change this to just `ObjKey altnernative_key = k;`, all the tests pass. Was there a reason we need the alternative key to be unresolved?

Reply:
Looking at this deeper, inside the loop we add 1 to the unresolved key which turns it from `(-2 - (k.value))` (i.e. unresolved) into `(-1 - (k.value))` which is just invalid. I think if you want these to be unresolved this way, you'll need to keep a counter around and increment the value of `k` and turn it into an unresolved key for each iteration of the loop.

Reply:
This can all be fixed by subtracting 1 from the alternative key instead of adding.